cli/data.go

```diff
@@ -601,118 +656,151 @@ func filenameForDownload(meta *datapb.BinaryMetadata) string {
+func (c *viamClient) tabularData(dest string, request *datapb.ExportTabularDataRequest) error {
 	if err := c.ensureLoggedIn(); err != nil {
 		return err
 	}
 
-	if err := makeDestinationDirs(dst); err != nil {
+	if err := makeDestinationDirs(dest); err != nil {
 		return errors.Wrapf(err, "could not create destination directories")
 	}
 
-	var err error
-	var resp *datapb.TabularDataByFilterResponse
-	// TODO(DATA-640): Support export in additional formats.
-	//nolint:gosec
-	dataFile, err := os.Create(filepath.Join(dst, dataDir, "data.ndjson"))
-	if err != nil {
-		return errors.Wrapf(err, "could not create data file")
-	}
-	w := bufio.NewWriter(dataFile)
+	fmt.Fprintf(c.c.App.Writer, "Downloading..") //nolint:errcheck
 
-	fmt.Fprintf(c.c.App.Writer, "Downloading..") //nolint:errcheck // no newline
-	var last string
-	mdIndexes := make(map[string]int)
-	mdIndex := 0
-	for {
-		for count := 0; count < maxRetryCount; count++ {
-			resp, err = c.dataClient.TabularDataByFilter(context.Background(), &datapb.TabularDataByFilterRequest{
-				DataRequest: &datapb.DataRequest{
-					Filter: filter,
-					Limit:  uint64(limit),
-					Last:   last,
-				},
-				CountOnly: false,
-			})
-			fmt.Fprintf(c.c.App.Writer, ".") //nolint:errcheck // no newline
-			if err == nil {
-				break
+	for count := 0; count < maxRetryCount; count++ {
+		err := func() error {
+			dataFilePath := filepath.Join(dest, dataFileName)
+			dataFile, err := os.Create(dataFilePath) //nolint:gosec
+			if err != nil {
+				return errors.Wrapf(err, "could not create data file")
 			}
-		}
-		if err != nil {
-			return err
-		}
 
-		last = resp.GetLast()
-		mds := resp.GetMetadata()
-		if len(mds) == 0 {
-			break
-		}
-		// Map the current response's metadata indexes to those combined across all responses.
-		localToGlobalMDIndex := make(map[int]int)
-		for i, md := range mds {
-			currMDIndex, ok := mdIndexes[md.String()]
-			if ok {
-				localToGlobalMDIndex[i] = currMDIndex
-				continue // Already have this metadata file, so skip creating it again.
-			}
-			mdIndexes[md.String()] = mdIndex
-			localToGlobalMDIndex[i] = mdIndex
+			writer := bufio.NewWriter(dataFile)
+			numWrites := uint64(0)
 
-			mdJSONBytes, err := protojson.Marshal(md)
-			if err != nil {
-				return errors.Wrap(err, "could not marshal metadata")
-			}
-			//nolint:gosec
-			mdFile, err := os.Create(filepath.Join(dst, metadataDir, strconv.Itoa(mdIndex)+".json"))
-			if err != nil {
-				return errors.Wrapf(err, fmt.Sprintf("could not create metadata file for metadata index %d", mdIndex)) //nolint:govet
-			}
-			if _, err := mdFile.Write(mdJSONBytes); err != nil {
-				return errors.Wrapf(err, "could not write to metadata file %s", mdFile.Name())
-			}
-			if err := mdFile.Close(); err != nil {
-				return errors.Wrapf(err, "could not close metadata file %s", mdFile.Name())
-			}
-			mdIndex++
-		}
+			dataRowChan := make(chan []byte)
```

Comment:
What are your thoughts on making this a buffered channel? The reason im asking is lets say disk IO is slow, we can fetch more requests while waiting on that. (I realize that disk io will probably never be the limiting factor, but curious what your thoughts are.

Reply:
If disk IO is slow, we'll be constrained by that anyways. If we run into problems, perhaps we can explore doing something, but I'm of the mind to leave this as-is for now